// @nomiclabs/hardhat-ethers/internal/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("hardhat/config");
const plugins_1 = require("hardhat/plugins");
const custom_ethers_provider_1 = require("./custom-ethers-provider");
const helpers_1 = require("./helpers");
require("./type-extensions");
(0, config_1.extendEnvironment)((hre) => {
    hre.ethers = (0, plugins_1.lazyObject)(() => {
        const { ethers } = require("ethers");
        const provider = new custom_ethers_provider_1.CustomEthersProvider(hre.network.provider, hre.network.name);
        return {
            ...ethers,
            provider,
            getSigner: (address) => (0, helpers_1.getSigner)(hre, address),
            getSigners: () => (0, helpers_1.getSigners)(hre),
            getImpersonatedSigner: (address) => (0, helpers_1.getImpersonatedSigner)(hre, address),
            // We cast to any here as we hit a limitation of Function#bind and
            // overloads. See: https://github.com/microsoft/TypeScript/issues/28582
            getContractFactory: helpers_1.getContractFactory.bind(null, hre),
            getContractFactoryFromArtifact: helpers_1.getContractFactoryFromArtifact.bind(null, hre),
            getContractAt: helpers_1.getContractAt.bind(null, hre),
            getContractAtFromArtifact: helpers_1.getContractAtFromArtifact.bind(null, hre),
            deployContract: helpers_1.deployContract.bind(null, hre),
        };
    });
});
